### PR TITLE
Add contextual retries

### PR DESCRIPTION
**Pull request by erinpentecost**
Added go mod and another method that doesn't throw away so many errors and allows you to cancel via contexts.

### CODE REVIEW

File: retry.go [hunk 1]
```diff
@@ -47,6 +49,42 @@ func X(x int, maxBackoff time.Duration, f func() bool) {
...
+// are done. Calls to f have a sleep duration between them.
+// XWithError will return a wrapped error around f's errors
+// if all attempts fail.
+// The attempts can be cancelled with ctx. If f does not cancel
+// when ctx is done, then the currently-running f will be allowed
+// to complete first.
+//
+// Example 1:
+//    var err error
+//    retry.XWithError(3, 5*time.Second, func() error {
```
Comment (epsniff) [inline retry.go]: copy pasta?

File: retry.go [hunk 2]
```diff
@@ -47,6 +49,42 @@ func X(x int, maxBackoff time.Duration, f func() bool) {
 	}
 }
 
+// XWithContext runs function f until f returns nil or the
+// number of retries exceeds x. Never more than x calls of f
+// are done. Calls to f have a sleep duration between them.
+// XWithError will return a wrapped error around f's errors
```
Comment (epsniff) [inline retry.go]: copy pasta?

**Review by epsniff — APPROVED** [review]

File: retry.go (hunk 2 above)
Comment (erinpentecost) [reply retry.go]: I forgot to change the comment after changing the name. Had a hard time naming it.

File: retry.go [hunk 3]
```diff
@@ -47,6 +49,42 @@ func X(x int, maxBackoff time.Duration, f func() bool) {
...
+// Example 1:
+//    var err error
+//    retry.XWithContext(context.Background(), 3, 5*time.Second, func(ctx context.Context) error {
+//        err = DoSomething()
+//        return err
+//    })
+func XWithContext(ctx context.Context, x int, maxBackoff time.Duration, f func(ctx context.Context) error) error {
+	var latestErr error
+	for i := 0; i < x; i++ {
+		timer := time.NewTimer(backoff(i, maxBackoff))
```
Comment (ajroetker) [inline retry.go]: Can we move this out of the for loop and just use reset?

File: retry.go [hunk 4]
```diff
@@ -47,6 +49,42 @@ func X(x int, maxBackoff time.Duration, f func() bool) {
...
+			return ctx.Err()
+		case <-timer.C:
+			if latestErr = f(ctx); latestErr == nil {
+				// finished ok!
+				return nil
+			}
+		}
+	}
+	// ran out of retries
+	return fmt.Errorf("all attempts failed, last attempt: %w", latestErr)
```
Comment (ajroetker) [inline retry.go]: Can we not wrap the error with anymore text? We use this all over the place

File: retry.go [hunk 5]
```diff
@@ -47,6 +49,42 @@ func X(x int, maxBackoff time.Duration, f func() bool) {
 	}
 }
```
Comment (epsniff) [inline retry.go:51]: FYI in those places where we wanted the ability to break a retry loop early, we use this package expbackoff.Backoff(try, 1, 10, time.Second)

API wise, I like the one in this PR better.

File: retry.go (hunk 3 above)
Comment (epsniff) [reply retry.go]: Good call, quickly creating Timers can cause lots of heap pressure.

Comment (erinpentecost) [reply retry.go]: Moved it out. It gets a little funky when you reset twice, so there's still a conditional in the loop for reset/new.

File: retry.go (hunk 5 above)
Comment (erinpentecost) [reply retry.go:51]: Do you mean this retry.XWithContext is better or expbackoff is better?

**Review by epsniff — APPROVED** [review]

**Pull request merged**